engine/execution/ingestion/engine.go

```diff
@@ -927,344 +1006,380 @@ func (e *Engine) generateExecutionReceipt(
-		e.syncWg.Add(1)
-		go e.saveDelta(ctx, executionStateDelta)
+	log := e.log.With().
+		Hex("sender", originID[:]).
+		Uint64("sealed", sealedHeight).
+		Uint64("from", req.FromHeight).
+		Uint64("to", req.ToHeight).
+		Logger()
 
+	// ignore requests for unsealed height
+	if req.FromHeight > sealedHeight {
+		log.Info().Msg("receives state sync requests for unsealed height, ignore")
 		return nil
-	})
-}
-
-func (e *Engine) saveDelta(ctx context.Context, executionStateDelta *messages.ExecutionStateDelta) {
+	}
 
-	defer e.syncWg.Done()
+	// fromHeight, toHeight must be sealed height
+	fromHeight, toHeight := req.FromHeight, req.ToHeight
+	if toHeight > sealedHeight {
+		toHeight = sealedHeight
+	}
 
-	log := e.log.With().
-		Hex("block_id", logging.Entity(executionStateDelta.Block)).
-		Uint64("block_height", executionStateDelta.Block.Header.Height).
-		Logger()
+	// for each height starting from fromHeight to toHeight,
+	// query the statecommitment, and if exists, send the
+	// state delta
+	// TOOD: add context
+	ctx := context.Background()
+	err = e.deltaRange(ctx, fromHeight, toHeight,
+		func(delta *messages.ExecutionStateDelta) {
+			err := e.syncConduit.Submit(delta, originID)
+			if err != nil {
+				e.log.Error().Err(err).Msg("could not submit block delta")
+			}
+		})
 
-	// synchronize DB writing to avoid tx conflicts with multiple blocks arriving fast
-	err := e.blocks.Store(executionStateDelta.Block)
 	if err != nil {
-		// It's possible for the parent of the target block to have arrived already. Don't fail here
-		if !errors.Is(err, storage.ErrAlreadyExists) {
-			log.Fatal().
-				Err(err).Msg("could not store block from delta")
-		}
+		return fmt.Errorf("could not send deltas: %w", err)
 	}
 
-	for _, collection := range executionStateDelta.CompleteCollections {
-		collection := collection.Collection()
-		err := e.collections.Store(&collection)
+	log.Info().Msg("responded state deltas")
+
+	return nil
+}
+
+// deltaRange querys it's local execution state to find deltas for a height
+// range between the fromHeight to the toHeight. If delta is found, then
+// pass it to the onDelta callback.
+func (e *Engine) deltaRange(ctx context.Context, fromHeight uint64, toHeight uint64,
+	onDelta func(*messages.ExecutionStateDelta)) error {
+
+	for height := fromHeight; height <= toHeight; height++ {
+		header, err := e.state.AtHeight(height).Head()
 		if err != nil {
-			log.Fatal().
-				Err(err).Msg("could not store collection from delta")
+			return fmt.Errorf("could not query block header at height: %v", height)
+		}
+
+		blockID := header.ID()
+		_, err = e.execState.StateCommitmentByBlockID(ctx, blockID)
+
+		if err == nil {
+			// this block has been executed, we will send the delta
+			delta, err := e.execState.RetrieveStateDelta(ctx, blockID)
+			if err != nil {
+				return fmt.Errorf("could not retrieve state delta for block %v, %w", blockID, err)
+			}
+
+			onDelta(delta)
```

Comment:
Previously, we [hold all the deltas in memory before sending them](https://github.com/onflow/flow-go/pull/31/files#diff-1926b4fb106417b1a85fc921a7a05e4fL131-L141), this could cause the memory usage increase a lot if an execution is syncing from a very old state.

Now, we will send back deltas one by one and won't them in memory.